src/shape.rs

```diff
@@ -660,237 +676,215 @@ impl ShapeLine {
-                                number_of_blanks -= 1;
-                            } else if let Some(previous_word) = span.words.get(i - 1) {
-                                // Current word causing a wrap is not whitespace, so we ignore the
-                                // previous word if it's a whitespace
-                                if previous_word.blank {
-                                    number_of_blanks -= 1;
-                                    prev_word_width = Some(previous_word.x_advance * font_size);
+                            let mut trailing_space_width = None;
+                            if i > 0 {
+                                if let Some(previous_word) = span.words.get(i - 1) {
+                                    // Current word causing a wrap is not whitespace, so we ignore the
+                                    // previous word if it's a whitespace
+                                    if previous_word.blank {
+                                        trailing_space_width =
+                                            Some(previous_word.x_advance * font_size);
+                                        number_of_blanks = number_of_blanks.saturating_sub(1);
+                                    }
                                 }
                             }
-                            if let Some(width) = prev_word_width {
-                                word_ranges.push((
-                                    (i, 0),
+                            if let Some(width) = trailing_space_width {
+                                add_to_visual_line(
+                                    &mut current_visual_line,
+                                    span_index,
+                                    (i + 2, 0),
                                     fitting_start,
                                     word_range_width - width,
                                     number_of_blanks,
-                                ));
+                                );
                             } else {
-                                word_ranges.push((
+                                add_to_visual_line(
+                                    &mut current_visual_line,
+                                    span_index,
                                     (i + 1, 0),
                                     fitting_start,
                                     word_range_width,
                                     number_of_blanks,
-                                ));
+                                );
                             }
+                            visual_lines.push(current_visual_line);
+                            current_visual_line = VisualLine::default();
+
                             number_of_blanks = 0;
                             if word.blank {
                                 fit_x = line_width;
                                 word_range_width = 0.;
                                 fitting_start = (i + 1, 0);
```

Comment:
```suggestion
                                fitting_start = (i, 0);
```
In the congruent case, if the word is blank we don't include it in the next visual line by setting `fitting_start = (i + 1, 0)`, because the start index is inclusive. If the same should happen in the incongruent case, `fitting_start` should be set to `(i, 0)` because the end index is exclusive. With `i + 1` the blank word would still be included in the next visual line.

As a side note, does it make sense to rename `fitting_start` to `fitting_end` in the incongruent case because it will always be used as an end index?

Reply:
it should be `i`. I've been trying to fix the trailing space that falls between the transition between LTR+RTL texts. Some of these changes were for that. But I haven't figured it out yet.
Like this space:

![Screenshot from 2023-03-13 12-59-35](https://user-images.githubusercontent.com/2222199/224803170-c702eee7-3808-4674-8d5a-e3d64e2fa53e.png)